### ISSUE TITLE
Doesn't work with accounts using MFA

### ISSUE THREAD

**Issue by adamnovak** [aws/amazon-genomics-cli#124]
[//]: #  (Please do not include your AWS account or credentials as part of the bug request.)

**Describe the Bug**

[//]: #  (Please describe the behavior that you are seeing with as many details as possible including logs from AGC, CloudWatch snippets and CloudFormation errors if relevant.)

I can't activate AGC on my AWS account because the tool can't work with the MFA assume role setup that we use for access control at the UCSC Genomics Institute.

**Steps to Reproduce**

[//]: #  (Include a list of steps that can be used to reproduce the bug.)

1. Set up your AWS accounts. Have one AWS account where all the users are, and get an access key and secret key for your user, and put them in `~/.aws/credentials`. Have another AWS account that is the one you actually want to run AGC in. Make a role there to be assumed, and set up the IAM permissions so that users in the first account can assume the role in the second account when authenticating with MFA. Get the ARN for the role.
2. Set up an MFA device on your IAM user in the first account and get its serial.
3. Write a `~/.aws/config` file that looks something like this, defining a profile that assumes the role in the destination account:
```
[default]
region = us-west-2

[profile toil]
region = us-west-2
role_arn = PASTE_ARN_HERE
source_profile = default
mfa_serial = PASTE_SERIAL_HERE
```
4. Try and set up AGC using the profile you defined:
```
agc account -v -p toil activate
```

**Relevant Logs**

[//]: #  (The logs that where obtained by running the commands here. Please run the command with "-v" so that we can see the verbose logs.)

```
$ agc account -v -p toil activate
2021-10-27T09:48:17-07:00 ☠️   error="assume role with MFA enabled, but AssumeRoleTokenProvider session option not set."
```

**Expected Behavior**

[//]: #  (A clear and concise description of what you expected to happen.)

The activate command should succeed, or else complain that the role doesn't have the permissions it needs to do the setup.

**Actual Behavior**

[//]: #  (A clear and concise description of what actually happened.)

Instead, AGC can't even assume the role; it looks like the piece of the Go SDK that is responsible for interacting with the keyboard and prompting for an MFA code is not set up. Interestingly, the exact same thing happens even when I have an active cached MFA session for the `aws` Python-based tool, so AGC isn't even capable of finding and using the cached assumed-role credentials in `~/.aws/cli/cache/`.

**Screenshots**

[//]: #  (If applicable, add screenshots to help explain your problem.)

Not applicable; everything is text.

**Additional Context**

[//]: #  (Add any other context about the problem here. The minimum required information is below)

Operating System: Linux courtyard 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 GNU/Linuxa
AGC Version: `1.0.0-41-gc5ac696`
Was AGC setup with a custom bucket: Not set up yet
Was AGC setup with a custom VPC: Not set up yet

**Comment by eddie1183** [comment]
Hi  adamnovak,

AGC currently relies on CDK for authenticated AWS calls. CDK currently does not support MFA, see  https://github.com/aws/aws-cdk/issues/10867. Once the CDK team adds MFA support we will be able to add MFA support to AGC.

Thanks,
Eddie

**Comment by adamnovak** [comment]
It looks like the CDK caching issue really goes all the way back to the JS SDK; there was a cache implementation for it in https://github.com/aws/aws-sdk-js/pull/2270 that was rejected, apparently due to lack of prompt interest on the AWS side.

But even without caching, the CDK does let you do MFA if you enter the MFA token each time; it looks like the real blocker here is AGC's Go code not at least supporting that. Once that's in, then we'd need to make sure that Go and the CDK were all talking to the same credential cache, to stop you from having to enter the MFA codes repeatedly.

To make this begin to work, the code here:

https://github.com/aws/amazon-genomics-cli/blob/457f231ce73c98f9a1a72bed80e2ddb1338cd862/packages/cli/internal/pkg/aws/provider.go#L148-L152

needs to do something like:

https://github.com/aws/aws-sdk-go-v2/blob/48e0568fe185ff56f68d1a11120170776b24394a/credentials/stscreds/assume_role_provider.go#L68-L99

Unfortunately the Go SDK folks don't give an example of how to assume roles *when specified by the config file*, using a token provider. We need to figure out how to get the provider into the config objects' redential resolution logic. Here is where it determines that it needs to use an MFA serial:

https://github.com/aws/aws-sdk-go-v2/blob/f5b68d198dbefe144696aa0051ed854571b5d614/config/resolve_credentials.go#L401-L404

Then it has a hook where it calls `getAssumeRoleCredentialProviderOptions` to get another function that gets to modify the `AssumeRoleOptions` object, and then by here we need a token provider set:

https://github.com/aws/aws-sdk-go-v2/blob/f5b68d198dbefe144696aa0051ed854571b5d614/config/resolve_credentials.go#L423-L425

It looks like we go through a few providers and things, and eventually the hook is brought out to the `AssumeRoleCredentialOptions` field of the `LoadOptions` instance:

https://github.com/aws/aws-sdk-go-v2/blob/b7d8e15425d2f86a0596e8d7db2e33bf382a21dd/config/load_options.go#L110-L112

Which is an object that you get to modify in a callback you pass to `LoadDefaultConfig`:

https://github.com/aws/aws-sdk-go-v2/blob/b7d8e15425d2f86a0596e8d7db2e33bf382a21dd/config/config.go#L170

It looks like the SDK already provides a helper we can use to copy-paste in some `AssumeRoleCredentialsOptions`:

https://github.com/aws/aws-sdk-go-v2/blob/c52469963c849e616df561b1110442f793ac3228/config/load_options.go#L452-L463

I think we just have to pass that along when setting up the configs, with options mentioning the `StdinTokenProvider`.

**Comment by adamnovak** [comment]
This is still an issue; there's a fix proposed in #231. I might be the one responsible for rebasing it actually.
